File: examples/cpp/example_publisher.cc
//
// Created by pairs on 7/26/18.
//

#include <test_message.pb.h>
#include <zros/zros.h>
#include <zros/node_handle.h>

int main() {

    // 1. init & run service discovery
    // for service discovery
    const std::string master_address = "localhost:23333";
    zros::init(master_address);

    // 2. create node
    const std::string node_name = "example_publisher";
    auto node_handle = std::make_shared<zros::NodeHandle>(node_name);

    // 3. create publisher
    const std::string topic = "test_topic";
    auto pub = node_handle->advertise<zros_example::TestMessage> (topic);

    std::thread pub_thread = std::thread([pub](){
        while(true) {
            zros_example::TestMessage message;
            message.set_detail("hello world");
            pub->publish(message);
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
    });
    pub_thread.join();
}

File: examples/python/example_publisher.py
# -*- coding: utf-8 -*-
""" An example for publisher

data: 2018-08-02
author: pairs
"""

import zros_python.zros as zros
import zros_python.node_handle as node
import zros_python.pb.test_message_pb2 as tmpb2
import time
import traceback
import logging

logger = logging.getLogger(__name__)

_PUB_TICK_TIME = 1

def publish():
    # 1. init & run service discovery
    master_address = u'localhost:23333'
    zros.init(master_address)

    # 2. create node
    node_name = u'example_publisher'
    node_handle = node.NodeHandle(node_name)

    # 3. create pub
    topic = u'test_topic'
    pub = node_handle.advertise(topic, tmpb2.TestMessage)
    if pub is None:
        logger.error(u'publisher %s advertise failed', topic)
        return

    # 4. publish message
    message = tmpb2.TestMessage()
    message.detail = u'hello world'
    message.count = 7
    while True:
        try:
            pub.publish(message)
        except Exception:
            logger.error(traceback.format_exc())
        time.sleep(_PUB_TICK_TIME)


if __name__ == '__main__':
    logging.basicConfig(format=u'[%(name)s-%(levelname)s]: %(message)s',
                        level=logging.NOTSET)
    publish()
